src/platforms/esp32/components/avm_sys/smp.c

```diff
@@ -0,0 +1,220 @@
+{
+    pthread_mutex_t mutex;
+};
+
+struct CondVar
+{
+    pthread_cond_t condvar;
+};
+
+struct RWLock
+{
+#ifdef HAVE_PTHREAD_RWLOCK
+    pthread_rwlock_t lock;
+#else
+    pthread_mutex_t lock;
+#endif
+};
+
+// ESP-IDF SDK explicitly mentions support for C11 with "__thread" keyword
+static __thread bool g_sub_main_thread = false;
+static uint32_t _Atomic g_pinned_cores = 0x1;
+
+static void *scheduler_thread_entry_point(void *arg)
+{
+    g_sub_main_thread = true;
+    void* result = (void *) (uintptr_t) scheduler_entry_point((GlobalContext *) arg);
+    BaseType_t core = xTaskGetAffinity(NULL);
+    if (core != -1) {
+        uint32_t desired = 1;
+        uint32_t expected = 3;
+        do {
+            desired = expected & ~(1 << core);
+        } while (!atomic_compare_exchange_weak(&g_pinned_cores, &expected, desired));
+    }
+    return result;
+}
+
+void smp_scheduler_start(GlobalContext *ctx)
+{
+    // When this function is called, the schedulers mutex is held and
+    // therefore it is safe to call esp_pthread_set_default_config
+    // g_pinned_cores is atomic in case another thread is being stopped.
+    esp_pthread_cfg_t esp_pthread_cfg = esp_pthread_get_default_config();
+    if (esp_pthread_cfg.pin_to_core != tskNO_AFFINITY) {
+        int core;
+        uint32_t expected = 1;
+        uint32_t desired;
+        do {
+            core = 1;
+            while (expected & 1 << core) {
+                core++;
+            }
+            desired = expected | (1 << core);
+        } while (!atomic_compare_exchange_weak(&g_pinned_cores, &expected, desired));
+
+        esp_pthread_cfg.pin_to_core = core;
+        if (UNLIKELY(esp_pthread_set_cfg(&esp_pthread_cfg))) {
+            AVM_ABORT();
+        }
+    }
+    pthread_t thread;
+    pthread_create(&thread, NULL, scheduler_thread_entry_point, ctx);
+    pthread_detach(thread);
+}
+
+bool smp_is_main_thread(GlobalContext *glb)
+{
+    UNUSED(glb);
+    return !g_sub_main_thread;
+}
+
+Mutex *smp_mutex_create()
+{
+    Mutex *result = malloc(sizeof(Mutex));
+    pthread_mutex_init(&result->mutex, NULL);
+    return result;
+}
+
+void smp_mutex_destroy(Mutex *mtx)
+{
+    pthread_mutex_destroy(&mtx->mutex);
+    free(mtx);
+}
+
+void smp_mutex_lock(Mutex *mtx)
+{
+    pthread_mutex_lock(&mtx->mutex);
+}
+
+bool smp_mutex_trylock(Mutex *mtx)
+{
+    int r = pthread_mutex_trylock(&mtx->mutex);
+    return r == 0;
+}
+
+void smp_mutex_unlock(Mutex *mtx)
+{
+    pthread_mutex_unlock(&mtx->mutex);
+}
+
+CondVar *smp_condvar_create()
+{
+    CondVar *result = malloc(sizeof(CondVar));
+    pthread_cond_init(&result->condvar, NULL);
+    return result;
+}
+
+void smp_condvar_destroy(CondVar *cv)
+{
+    pthread_cond_destroy(&cv->condvar);
+    free(cv);
+}
+
+void smp_condvar_wait(CondVar *cv, Mutex *mtx)
+{
+    pthread_cond_wait(&cv->condvar, &mtx->mutex);
+}
+
+void smp_condvar_signal(CondVar *cv)
+{
+    pthread_cond_signal(&cv->condvar);
+}
+
+RWLock *smp_rwlock_create()
```

Comment:
I'm not sure about this semantic:
- dynamic memory allocation might fail, so we should do a number of checks around
- memory allocation has a certain memory overhead over simple embedding of RWLock struct

I rather suggest `smp_rwlock_init(RWLock *lock)` rather than returning a new one.
If you agree with this different approach, I realize that amending this might be too complex, so adding an additional commit on top of the other changes is fine for me.

Reply:
The reason I went with this design is the size of smp structures (threads, mutexes, locks) is platform dependent. We cannot include pthread.h as it doesn't exist on some platforms we may want SMP for (Pico).

Reply:
Resolved as I added missing checks.
Few locks are created (this doesn't depend on the processes), so I think the overhead should be fine. Feel free to unresolve if you have a better design in mind.